processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -0,0 +1,195 @@
+				delete(nameToMetricMapping, transform.MetricName)
+			} else if transform.Action == Insert {
+				var newMetric *metricspb.Metric
+				mds[i].Metrics, newMetric = mtp.insert(metric, mds[i].Metrics, transform)
+				// mapping has to be updated with the name metric
+				nameToMetricMapping[newMetric.MetricDescriptor.Name] = newMetric
+			}
+		}
+	}
+
+	return pdatautil.MetricsFromMetricsData(mds)
+}
+
+// update updates the original metric content in the metricPtr pointer.
+func (mtp *metricsTransformProcessor) update(metricPtr *metricspb.Metric, transform Transform) {
+	// metric name update
+	if transform.NewName != "" {
+		metricPtr.MetricDescriptor.Name = transform.NewName
+	}
+
+	for _, op := range transform.Operations {
+		// update label
+		if op.Action == UpdateLabel {
+			mtp.updateLabelOp(metricPtr, op)
+		}
+		// aggregate across labels
+
+		// aggregate across label values
+
+	}
+}
+
+// insert inserts a new copy of the metricPtr content into the metricPtrs slice.
+// returns the new metrics list and the new metric
+func (mtp *metricsTransformProcessor) insert(metricPtr *metricspb.Metric, metricPtrs []*metricspb.Metric, transform Transform) ([]*metricspb.Metric, *metricspb.Metric) {
+	metricCopy := mtp.createCopy(metricPtr)
+	mtp.update(metricCopy, transform)
+	return append(metricPtrs, metricCopy), metricCopy
+}
+
+func (mtp *metricsTransformProcessor) updateLabelOp(metricPtr *metricspb.Metric, op Operation) {
+	// if new_label is invalid, skip this operation
+	if !mtp.validNewLabel(metricPtr.MetricDescriptor.LabelKeys, op.NewLabel) {
+		log.Printf("error running %q processor due to collided %q: %v with existing label on metric named: %v detected by function %q", typeStr, NewLabelFieldName, op.NewLabel, metricPtr.MetricDescriptor.Name, validNewLabelFuncName)
+		return
+	}
+
+	for _, label := range metricPtr.MetricDescriptor.LabelKeys {
+		if label.Key != op.Label {
+			continue
+		}
+		// label key update
+		if op.NewLabel != "" {
+			label.Key = op.NewLabel
+		}
+		// label value update
+	}
+}
+
+// createCopy creates a new copy of the input metric.
+func (mtp *metricsTransformProcessor) createCopy(metricPtr *metricspb.Metric) *metricspb.Metric {
+	copyMetricDescriptor := *metricPtr.MetricDescriptor
+	copyLabelKeys := make([]*metricspb.LabelKey, 0)
+	for _, labelKey := range copyMetricDescriptor.LabelKeys {
+		copyLabelKeys = append(
+			copyLabelKeys,
+			&metricspb.LabelKey{
+				Key:         labelKey.Key,
+				Description: labelKey.Description,
+			},
+		)
+	}
+	copyMetricDescriptor.LabelKeys = copyLabelKeys
+
+	copy := &metricspb.Metric{
+		MetricDescriptor: &copyMetricDescriptor,
+		Timeseries:       metricPtr.Timeseries,
+		Resource:         metricPtr.Resource,
+	}
+	return copy
+}
+
+// validNewName determines if the new name is a valid one. An invalid one is one that already exists.
+func (mtp *metricsTransformProcessor) validNewName(transform Transform, nameToMetricMapping map[string]*metricspb.Metric) bool {
```

Comment:
If you do agree with me about removing this validation, you can remove a huge amount of the code in the core functions of this file, i.e. the `transform` function becomes simply:

```go
for _, metric := range data.Metrics {
    transform := nameToTransformMapping[metric.MetricDescriptor.Name]

    if transform.Action == Insert {
        metric = mtp.createCopy(metric)
        mds[i].Metrics = append(metricPtrs, metricCopy)
    }

    mtp.update(metricCopy, transform)
}
```

The `nameToTransformMapping` can be made once on initialization

Reply:
So this validNewName and other validation functions came out of discussions with Dave and Quentin. Basically, when these unexpected behaviors occur, not only does it appear in the output data, but there will be an error message logged, so that the errors can be more viewable to the users

Reply:
Also regarding simplifying the code this way, since we want the transformations to perform in the order from the list. I would think we need to iterate through the transforms instead of the metric, which then involves managing the metrics mapping as we go though transformations.

Reply:
Yea okay that's true. I was hoping to find a solution that would avoid having to allocate an extra map on each call to `ConsumeMetrics`, but if we want ordered operations that is probably not possible